add mark-as-done

tharvik opened the pull request:
## Checklist

- [x] I read and understood the **[Contributing Guidelines](https://github.com/Abrynos/ShoppingList/blob/main/.github/CONTRIBUTING.md)**.
- [x] This is not a **[duplicate](https://github.com/Abrynos/ShoppingList/pulls)** of an existing merge request.
- [x] This is not a **[translation update](https://crowdin.com/project/abrynosshoppinglist)**.
- [x] I believe this falls into the scope of the project and should be part of the built-in functionality.
- [x] My code follows the **[code style](https://github.com/Abrynos/ShoppingList/blob/main/.github/CONTRIBUTING.md#code-style)** of this project.
- [x] All new and existing tests pass.

## Changes

* update `ShoppingListAdapter.changeItemCallback` to use `null` for deletion instead of zero
* fade-out substract button when amount equals zero

### New functionality

fade-out name when amount equals zero

### Changed & removed functionality

substracting when zero no longer delete item. it pushes you to use the dedicated button to do that (wanted to add a toaster notification but felt a bit out of scope).

Abrynos commented:
As per [my comment](https://github.com/Abrynos/ShoppingList/issues/106#issuecomment-2084393024) on issue #106 :

> please make sure it's behind a new setting

Abrynos commented on app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/adapters/ShoppingListAdapter.kt:
```diff
@@ -52,15 +52,17 @@ class ShoppingListAdapter(private val changeAmountCallback: BiConsumer<RequiredI
 		binding.amountTextView.isVisible = item.amount > 1
 
 		binding.nameTextView.text = item.name
+		binding.nameTextView.alpha = if (item.amount > 0) 1f else 0.3f
```
The alpha value should be a constant value since it's used more than once.
```kotlin
companion object {
    private const val ZERO_COUNT_OPACITY = 0.3f
}
```

tharvik commented:
> As per [my comment](https://github.com/Abrynos/ShoppingList/issues/106#issuecomment-2084393024) on issue #106 :

yeah, it's kinda hard to know which issues are relevant to a wanted features as all are closed. consider keeping some open, eventough you don't intend to work on it personnally, as to avoid newcomers opening the same issue (#40, #44, #51, #55, ~#75, #106 in this case) and external contributors to have a clearer summary of what is wanted.

> please make sure it's behind a new setting

added. it's a bit weird to handle both use-cases, as two buttons have the same action on an item with an amount of one (minus and delete).
also, when changing the crossout setting, I didn't find a nice way to refresh the presentation of the items in the main activity so until the amount change, it keeps the previous style. maybe you have an idea?

Abrynos commented:
> when changing the crossout setting, I didn't find a nice way to refresh the presentation of the items in the main activity so until the amount change, it keeps the previous style. maybe you have an idea?

You just gave me a headache because that reminded me of the colour-switch and all the brain-cells I lost making that work. I just changed my mind about having it as a setting (really sorry for the extra work). Feel free to remove that again.

Abrynos commented:
However, it would be nice to have a boolean-setting named "Order zero-count items last" in order to move them to the bottom of the list. Boolean because non-zero items should still follow default sorting order.

tharvik commented:
> > when changing the crossout setting
>
> You just gave me a headache because that reminded me of the colour-switch and all the brain-cells I lost making that work. I just changed my mind about having it as a setting (really sorry for the extra work). Feel free to remove that again.

haa yes, android is quite cumbersome to work with. I removed it or rather, I changed it into a boolean switch for sorting zero-count items to the bottom, as you suggested. as such, fadeout of zero-count items name (and minus-button) is always on.
and as the sorting option wasn't actually, well, sorting, so I implemented it.

Abrynos commented on app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/activities/MainActivity.kt:
```diff
@@ -85,21 +86,44 @@ class MainActivity : AbstractShoppingActivity() {
+			Order.Unordered -> Comparator { _, _ -> 0 }
+			Order.Alphabetical -> compareBy { it.name }
+			Order.AmountAscending -> compareBy { it.amount }
+			Order.AmountDescending -> compareBy<RequiredItem> { it.amount }.reversed()
```
This should use `compareByDescending`. You also correctly noticed that we explicitly need to set the generic type of the comparator and I'd much prefer if it was on the first entry of the switch.

```suggestion
			Order.AmountDescending -> compareByDescending { it.amount }
```

tharvik replied:
indeed, that's more readable.

Abrynos commented on app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/activities/MainActivity.kt:
```diff
@@ -85,21 +86,44 @@ class MainActivity : AbstractShoppingActivity() {
+			else 0
+		}
+		val orderComparator = when (this.settings.order) {
+			Order.Unordered -> Comparator { _, _ -> 0 }
```
```suggestion
			Order.Unordered -> Comparator<RequiredItem> { _, _ -> 0 }
```

Abrynos commented on app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/activities/MainActivity.kt:
```diff
@@ -85,21 +86,44 @@ class MainActivity : AbstractShoppingActivity() {
 
 	private fun observeDatabaseChange(items: List<RequiredItem>) {
-		(binding.mainListRecyclerView.adapter as ShoppingListAdapter).setItems(items)
+		val zeroItemsComparator = Comparator<RequiredItem> { a, b ->
```
We should probably only initialize the `zeroItemsComparator` in case the setting is active. Move it under the `if` below.

tharvik replied:
moved it to an `if` which returns the chained `Comparator`s. the else branch feels a bit akward but that's just me.

Abrynos replied:
Instead of doing the following:
```kotlin
val orderComparator = CREATE_ORDER_COMPARATOR
val comparator = if (this.settings.orderZeroItemsLast) {
    val zeroComparator = CREATE_ZERO_COMPARATOR
    zeroComparator.then(orderComparator)
} else {
    orderComparator
}
val sorted = items.sortedWith(comparator)
```
we can do this:
```kotlin
var comparator = CREATE_ORDER_COMPARATOR
if (this.settings.orderZeroItemsLast) {
    comparator = CREATE_ZERO_COMPARATOR.then(comparator)
}
val sorted = items.sortedWith(comparator)
```

Abrynos replied:
Also we don't need the `this` there.

tharvik replied:
ho, good idea, I'm much into functional so having mutable variables are really not into my realm. thanks for the hint.

Abrynos commented on app/src/main/res/layout/activity_options.xml:
```diff
@@ -83,6 +84,13 @@
+		android:onClick="onClickOrderZeroItemsLastSwitch"
+		android:text="@string/optionsOrderZeroItemsLast"
+		android:textSize="@dimen/optionTextSize"
+		app:layout_constraintTop_toBottomOf="@id/optionsOrderDropdown" />
```
> This view is not constrained horizontally

As with the one at the very top we can ignore the warning (or constrain it to start and end of parent).
```suggestion
		app:layout_constraintTop_toBottomOf="@id/optionsOrderDropdown"
		tools:ignore="MissingConstraints" />
t" />
```

tharvik replied:
ho right, good catch, I forgot to run the linter!

Abrynos commented on app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/settings/Settings.kt:
```diff
@@ -41,6 +41,9 @@ class Settings(context: Context) {
 		private val ORDER_DEFAULT: Order = Order.Unordered
+
+		private const val ORDER_ZERO_ITEMS_LAST_NAME = "orderZeroItemsLast"
+		private val ORDER_ZERO_ITEMS_LAST_DEFAULT = true
```
Can be `const`
```suggestion
		private const val ORDER_ZERO_ITEMS_LAST_DEFAULT = true
```

tharvik replied:
haa thanks, that's safer.
FWIW I'm not used to kotlin yet (but why enum can't be `const`?), it's actually the first time in writing some.

Abrynos replied:
> why enum can't be `const`?

I'm not a kotlin-pro either, but from what I understood enums in kotlin are just a special kind of class and not a numeric type. Since kotlin compiles down to code understandable by the JVM and it's not possible for class instances to be `const` there, enums can't be in kotlin. You may also notice that in the declaration at the bottom of the file in our case they even implement an interface (which I use to automatically handle both color and order dropdowns with only one implementation):

```kotlin
enum class AccentColor(val value: Int, val styleResourceId: Int, private val description: Int) : DescriptiveSetting {
```

tharvik replied:
right, the JVM is kinda old in this regard, as the enum are, as you said, classes and not discrete values of a common type. it might be enforcable at the kotlin compiler level but would be weird to interface with classic java enum. I think that Scala does it better, having sealed classes to work around it.
anyway, thanks for the explanation!

Abrynos reviewed (changes requested):
Otherwise I like where this is going.

Abrynos commented:
Also feel free to mention yourself in `fastlane/metadata/android/en-US/changelogs/17.txt` 😉

tharvik commented:
> Also feel free to mention yourself in `fastlane/metadata/android/en-US/changelogs/17.txt` 😉

haha, added in latest commit :)

Abrynos commented on app/src/main/res/layout/activity_options.xml:
```diff
@@ -83,6 +84,15 @@
 	</LinearLayout>
 
+	<androidx.appcompat.widget.SwitchCompat
+		android:id="@+id/optionsOrderZeroItemsLast"
```
On the other controls, the id refers to the type of control (e.g. `optionsThemeSwitch` or `optionsOrderDropdown`). For consistency this should probably be something like `optionsOrderZeroItemsLastSwitch` (long name, I know. Maybe you can think of something better)

tharvik replied:
long name yes but internal and only used a few time. I personnaly do not encode the type in the name, it's checked by the compiler and should be obvious by the context but changing all the other IDs here (to be consistent) is a bit out of place IMO.
renamed to end id with "Switch" and string with "Text".

Abrynos commented:
Thank you for your contribution 🏆
